hail/src/main/scala/is/hail/utils/NDArray.scala

```diff
@@ -0,0 +1,31 @@
+package is.hail.utils
+
+import org.apache.spark.sql.Row
+import org.json4s.{JArray, JInt, JObject, JValue}
+
+object NDArray {
+  def fromRow(x: Row): NDArray =
+    new NDArray(
+      x.getLong(0),
+      x.getAs[Array[Int]](1),
+      x.getLong(2),
+      x.getAs[Array[Int]](3),
+      x.getAs[Array[Any]](4))
+}
+
+class NDArray(
+  val flags:Long, 
+  val shape: Array[Int], 
+  val offset: Long, 
+  val strides: Array[Int], 
+  val data: Array[Any]) extends Serializable {
```

Comment:
I don't think we should do this.  We've learned that `Array[Any]` is not something we can do at scale, so this can only turn out to be a mistake.  Also, @catoverdrive has set a goal and has a plan to remove all Java values from pipeline execution, so this will all get deleted soon.

ndarrays should only be implemented in the C++ code generator for now.

As part of her plan, the JSONAnnotationImpex will get replaced by a RegionValue to JSON converter.  I think converting ndarray to JSON as a n-nested JSON array that can be fed directly into the numpy array constructor is probably the right thing.  Specifically, that means we don't kick out the strides and offsets (and data that is no longer being referenced in the case of slicing), just the values this ndarray represents.

The only reason we really convert to JSON is communicating between the Hail Python API and the Hail rest service.  Once we have support for Hail encoded region values on the Python side, that protocol can become binary and we won't particularly need ndarray to JSON support.

Reply:
this isn't intended to be anything but a serialization intermediate -- I think it makes sense for Jon to mirror the other type/value hierarchy and forge ahead.

```
Once we have support for Hail encoded region values on the Python side, that protocol can become binary
```

This doesn't seem like a particularly big deal at the moment - the python json deserialization overhead has been low historically. It will increase as we make the rest of the stack faster.

Reply:
What's a serialization intermediate?

Reply:
I think we are close to this state of affairs and I would prefer the effort be spent pushing us there rather than building on these deprecated interfaces that are slated for deletion.  In particular, it is straightforward to rewrite str and json to work on region values right now and there is nothing stopping that form happening.

Reply:
alright, I'm happy to be overruled. 

I suppose we can also just not support str/json on NDArray in the first few PRs, and just build up the IR infrastructure.

Reply:
Yeah, exactly.  We're still a ways from having a non-trivial runtime ndarray value.  When we get to the point of wanting to print them or send them over the service interface, I think we should just write non-annotation based str/json, since it needs to happen soon anyway.